src/99-appendices/15-inheritance-principle.md

```diff
@@ -0,0 +1,260 @@
+# Appendix XV: The Inheritance Principle
+
+The [Inheritance Principle](#inheritance-principle) specifies how it is
+not strictly necessary to have data and metadata files with identical names
+apart from the [file extension](#definitions).
+It is possible to have a single metadata file with its contents applicable
+to multiple data files;
+it is additionally possible to have one data file where there are multiple
+metadata files from which metadata are applicable.
+This Appendix provides the precise definition of how this principle operates
+and the consequences of such,
+as reference for both software implementations and advanced BIDS users.
+
+## Rules
+
+1.  Any metadata file (such as `.json`, `.bvec` or `.tsv`) MAY be defined at any directory level.
+
+1.  For a given data file, any metadata file is applicable to that data file if:
+    1.  It is stored at the same directory level or higher;
+    1.  The metadata and the data filenames possess the same suffix;
+    1.  The metadata filename does not include any entity absent from the data filename.
+
+1.  A metadata file MUST NOT have a filename that would be otherwise applicable
+    to some data file based on rules 2.b and 2.c but is made inapplicable based on its
+    location in the directory structure as per rule 2.a.
+
+1.  If, for a given data file, multiple metadata files satisfy criteria 2.a-c above:
+
+    1.  The set of applicable metadata files is ordered as follows.
+        Within each level of the filesystem hierarchy independently,
+        applicable files are ordered from fewest to most entities;
+        each subsequent filename MUST possess a set of entities that is a
+        strict superset of the previous filename.
+        These lists are concatenated in order of highest to lowest level in the
+        filesystem hierarchy.
```

Comment:
```suggestion
        Levels of the filesystem hierarchy considered from the highest (top directory of the dataset) to lowest (folder with data files for a subject/session) level.
```
- `list` provides a somewhat "implementation dependent" detail, and not even mentioned in prior sentences
- not sure if we should try to similarly dissolve `Set` and just say "Applicable metadata files ..."?

Also wanted to provide explicit hint on highest/lowest, since depending how you would look for filesystem tree, e.g. root of a tree in nature is at its lowest level, and you climb up to tree to get higher.

Reply:
I am very concerned about the proposed rule: 

> +Where more than one metadata file is applicable to a data file,
> +the contents of the file closest in filesystem position to the data file
> +(lowest in the directory hierarchy, most entities in the file name)
> +takes precedence.

I believe this is the **opposite** of how it currently is specified and how all of the tools work. 

Currently, a key in a JSON file at the top level cannot be overridden by keys at the lower level. 

If the proposed changes are to be made, I think this needs a much broader discussion by a broader audience. 
@sappelhoff @rwblair

Reply:
Yep, I used "list" to refer to the set of applicable metadata files within a single level of the filesystem hierarchy, and there is then an ordering of those lists, but failed to define such.

I've been constantly battling to find a reader-friendly way to convey the fact that there are two "sorting mechansims": firstly down the filesystem hierarchy, and secondly within individual directories. I was trying to convey the idea that you first do a scan through to find all applicable metadata files, and then perform a sort on them prior to loading their contents, as opposed to just traversing and loading as you go. 

> Also wanted to provide explicit hint on highest/lowest, since depending how you would look for filesystem tree, e.g. root of a tree in nature is at its lowest level, and you climb up to tree to get higher.

-   I think I've had this discussion once before w.r.t. BIDS: The convention around here was that root = highest and one traverses down, and with my first contribution I had it the other way around.
-   I don't know the extent to which this is a globally accepted standard; it doesn't appear in "common principles", and maybe it should.
-   I'm slightly hesitant to briefly provide exemplars of highest / lowest levels given that there is some chance of the depth of the filesystem tree in BIDS increasing; see https://github.com/bids-standard/bids-bep016/issues/50.

Reply:
> I've been constantly battling to find a reader-friendly way to convey the fact that there are two "sorting mechansims": firstly down the filesystem hierarchy, and secondly within individual directories.

Just an observation which might (or not) help out: **are there really multiple mechanisms or is it just manifestation of one?**  Due to the fact that sub- and ses- are also entities, and they are stripped off the filename as you go higher (closer to root) in the hierarchy -- sorting by number of entities generalizes across filesystem hierarchy.

Reply:
Maaaybe. As per:

> I'm slightly hesitant to briefly provide exemplars of highest / lowest levels given that there is some chance of the depth of the filesystem tree in BIDS increasing

, it would depend on whether or not such an expansion were to occur *and* whether all entities present at higher levels would be enforced to propagate down to all files at lower levels. It's possible that file names could start getting frustratingly long in such circumstances. So linking https://github.com/bids-standard/bids-bep016/issues/50 again.

I suppose my thinking was that:

-   If existing users are familiar with the directory-based inheritance, and that is a much simpler process to understand, then it should be presented as a two-stage process, such that only "power users" wishing to perform inheritance within a directory need concern themselves with this augmentation.
-   When it comes to software implementation, one is likely to perform the metadata inheritance as currently described: order within the highest filesystem level, then step down the filesystem, ordering within each level and then concatenating. It would not make sense to put all applicable files across all directories into a single list and then sort them, even if it would produce the same result.

I'll have a go at just improving the description, and we can re-assess.

Reply:
See what you think of https://github.com/bids-standard/bids-specification/pull/1003/commits/290de6ed595e971cd8701fc4f3a4644c5d69c11c.